packages/lodestar/src/chain/errors/attestationError.ts

```diff
@@ -0,0 +1,199 @@
+   */
+  ERR_INVALID_SUBNET_ID = "ERR_INVALID_SUBNET_ID",
+  /**
+   * The attestation failed the `state_processing` verification stage.
+   */
+  ERR_INVALID = "ERR_INVALID",
+  /**
+   * There was an error whilst processing the attestation. It is not known if it is valid or invalid.
+   */
+  ERR_BEACON_CHAIN_ERROR = "ERR_BEACON_CHAIN_ERROR",
+}
+
+export type AttestationErrorType =
+  | {
+      code: AttestationErrorCode.ERR_FUTURE_SLOT;
+      attestationSlot: Slot;
+      latestPermissibleSlot: Slot;
+    }
+  | {
+      code: AttestationErrorCode.ERR_PAST_SLOT;
+      attestationSlot: Slot;
+      earliestPermissibleSlot: Slot;
+    }
+  | {
+      code: AttestationErrorCode.ERR_EMPTY_AGGREGATION_BITFIELD;
+    }
+  | {
+      code: AttestationErrorCode.ERR_INVALID_SELECTION_PROOF;
+      aggregatorIndex: ValidatorIndex;
+    }
+  | {
+      code: AttestationErrorCode.ERR_AGGREGATOR_NOT_IN_COMMITTEE;
+      aggregatorIndex: ValidatorIndex;
+    }
+  | {
+      code: AttestationErrorCode.ERR_AGGREGATOR_PUBKEY_UNKNOWN;
+      aggregatorIndex: ValidatorIndex;
+    }
+  | {
+      code: AttestationErrorCode.ERR_ATTESTATION_ALREADY_KNOWN;
+      root: Uint8Array;
+    }
+  | {
+      code: AttestationErrorCode.ERR_AGGREGATE_ALREADY_KNOWN;
+      root: Uint8Array;
+    }
+  | {
+      code: AttestationErrorCode.ERR_AGGREGATOR_INDEX_TOO_HIGH;
+      aggregatorIndex: ValidatorIndex;
+    }
+  | {
+      code: AttestationErrorCode.ERR_UNKNOWN_HEAD_BLOCK;
+      beaconBlockRoot: Uint8Array;
+    }
+  | {
+      code: AttestationErrorCode.ERR_BAD_TARGET_EPOCH;
+    }
+  | {
+      code: AttestationErrorCode.ERR_UNKNOWN_TARGET_ROOT;
+      root: Uint8Array;
+    }
+  | {
+      code: AttestationErrorCode.ERR_INVALID_SIGNATURE;
+    }
+  | {
+      code: AttestationErrorCode.ERR_NO_COMMITTEE_FOR_SLOT_AND_INDEX;
+      slot: Slot;
+      index: CommitteeIndex;
+    }
+  | {
+      code: AttestationErrorCode.ERR_NOT_EXACTLY_ONE_AGGREGATION_BIT_SET;
+      numBits: number;
+    }
+  | {
+      code: AttestationErrorCode.ERR_PRIOR_ATTESTATION_KNOWN;
+      validatorIndex: ValidatorIndex;
+      epoch: Epoch;
+    }
+  | {
+      code: AttestationErrorCode.ERR_FUTURE_EPOCH;
+      attestationEpoch: Epoch;
+      currentEpoch: Epoch;
+    }
+  | {
+      code: AttestationErrorCode.ERR_PAST_EPOCH;
+      attestationEpoch: Epoch;
+      currentEpoch: Epoch;
+    }
+  | {
+      code: AttestationErrorCode.ERR_ATTESTS_TO_FUTURE_BLOCK;
+      block: Slot;
+      attestation: Slot;
+    }
+  | {
+      code: AttestationErrorCode.ERR_INVALID_SUBNET_ID;
+      received: number;
+      expected: number;
+    }
+  | {
+      code: AttestationErrorCode.ERR_INVALID;
+      error: Error;
+    }
+  | {
+      code: AttestationErrorCode.ERR_BEACON_CHAIN_ERROR;
+      error: Error;
+    };
+
+export class AttestationError extends Error {
+  public type: AttestationErrorType;
+
+  constructor(type: AttestationErrorType) {
+    super(type.code);
```

Comment:
We can do in another PR, but I feel like we should handle type serialization to message in constructor of error. Something like:
```typescript
super(this.getMessage(type));


private getMessage(type: AttestationErrorType) {
   let message = type.code;
   switch(type.code) {
      case something:
            message += " details: " + JSON.stringify(type.block)
   }
   return message;
}
```

Reply:
Yeah, we definitely need a way to get this metadata into a printable state for logging purposes.